pkg/database/sqlparser/migrator/db.go

```diff
@@ -25,204 +24,128 @@ import (
-func ClearSandbox() {
-	sandbox = nil
+func (mig *Migrator) ClearSandbox() {
+	mig.sandbox = nil
 }
 
-func SandBox() *gorm.DB {
-	if sandboxDSN == "" {
-		logrus.Fatalln("sandbox DSN is invalid")
+func (mig *Migrator) SandBox() *gorm.DB {
+	if mig.sandbox != nil {
+		return mig.sandbox
 	}
 
-	if sandbox == nil {
-		var (
-			open           *sql.DB
-			create         *sql.Stmt
-			err            error
-			createDatabase = "CREATE DATABASE IF NOT EXISTS " + database + " DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci"
-			dropDatabase   = "DROP SCHEMA IF EXISTS " + database
-		)
-		defer func() {
-			if open != nil {
-				_ = open.Close()
-			}
-		}()
-		for now := time.Now(); time.Since(now) < time.Second*150; time.Sleep(time.Second * 3) {
-			open, err = sql.Open("mysql", sandboxDSN)
-			if err != nil {
-				logrus.Infof("failed to connect to MySQL sandbox, DSN: %s, err: %v", sandboxDSN, err)
-				continue
-			}
-
-			drop, err := open.Prepare(dropDatabase)
-			if err != nil {
-				logrus.Warnf("failed to Prepare %s stmt: %v", dropDatabase, err)
-				continue
-			}
-			if _, err := drop.Exec(); err != nil {
-				logrus.Warnf("failed to Exec %s: %v", dropDatabase, err)
-			}
-
-			create, err = open.Prepare(createDatabase)
-			if err != nil {
-				logrus.Warnf("failed to Prepare %s stmt, err: %v", strconv.Quote(createDatabase), err)
-				continue
-			}
-
-			if _, err = create.Exec(); err != nil {
-				logrus.Fatalf("failed to Exec prepared %s stmt, err: %v", strconv.Quote(createDatabase), err)
-			}
-			break
+	var (
+		open           *sql.DB
+		create         *sql.Stmt
+		err            error
+		createDatabase = "CREATE DATABASE IF NOT EXISTS " + mig.sandboxSettings.Name + " DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci"
+		dropDatabase   = "DROP SCHEMA IF EXISTS " + mig.sandboxSettings.Name
+	)
+	defer func() {
+		if open != nil {
+			_ = open.Close()
 		}
+	}()
+
+	var (
+		timeout = time.Second * 150
+		dsn     = mig.SandboxParameters().Format(false)
+	)
+
+	for now := time.Now(); time.Since(now) < timeout; time.Sleep(time.Second * 3) {
+		open, err = sql.Open("mysql", dsn)
 		if err != nil {
-			logrus.Fatalf("failed to dial MySQL sandbox, DSN: %s, err: %v", sandboxDSN, err)
+			logrus.WithField("DSN", dsn).WithError(err).Infoln("failed to connect to sandbox")
+			continue
 		}
 
-		values := make(url.Values, 2)
-		values.Add("timeout", "150s")
-		values.Add("parseTime", "True")
-		sandboxDSN += database + "?" + values.Encode()
-		sandbox, err = gorm.Open(mysql.Open(sandboxDSN), &gorm.Config{DisableForeignKeyConstraintWhenMigrating: true})
+		drop, err := open.Prepare(dropDatabase)
 		if err != nil {
-			logrus.Fatalf("failed to open connection to sandbox, sandbox dsn: %s, err: %v", sandboxDSN, err)
+			logrus.Warnf("failed to Prepare %s stmt: %v, may sandbox is not working yet, wait it for %v seconds", dropDatabase, err, int(timeout.Seconds()-time.Since(now).Seconds()))
```

Comment:
Please use %d format, no int() conversion is required.

Reply:
Placeholder argument 'timeout.Seconds()-time.Since(now).Seconds()' has the wrong type 'float64' (%d)

Reply:
use %.1f ?

Reply:
resolved